teps/0075-object-param-and-result-types.md

```diff
@@ -0,0 +1,833 @@
+            * name: The name of the package
+            * version: The version of the package
+         schema:
+           type: object
+           properties:
+             name: {
+              type: string 
+             }
+             version: {
+              type: string 
+             }
+   ```
+   
+   Eventually when we have nested support, we could define a `wheel` interface which contains all of the above.
+
+2. Grouping related params to create simpler interfaces, and allowing similar Tasks to easily define pieces of their
+   interface. For example [the git-clone task has 15 parameters](https://github.com/tektoncd/catalog/blob/main/task/git-clone/0.4/README.md#parameters),
+   and [the git-rebase task has 10 parameters](https://github.com/tektoncd/catalog/blob/main/task/git-rebase/0.1/README.md#parameters).
+   Some observations:
+   * Each has potential groupings which stand out, for example:
+     * The git-rebase task could group the `PULL` and `PUSH` remote params; each object would need the same params,
+       which could become an interface for "git remotes"
+     * Potential groupings for the git-clone task stand out as well, for example the proxy configuration
+     * On that note, since the git-rebase task is using git and accessing remote repos, it probably needs the same
+       proxy configuration, so what they probably both need is some kind of interface for what values to provide when
+       accessing a remote git repo
+   Other examples include [okra-deploy](https://github.com/tektoncd/catalog/tree/main/task/orka-deploy/0.1) which is using
+   param name prefixes to group related params (e.g. `ssh-`, `okra-vm`, `okra-token`).
+
+## Requirements
+
+* Must be possible to programmatically determine the structure of the object
+* Must be possible for a result object to be empty
+* Must be possible to use the object results of one Task in a Pipeline as the
+  param of another Task in the pipeline which has a object param when the interface matches (more detail in
+  [required vs optional keys](#required-vs-optional-keys))
+* Must be possible to use one specific value in a object result of one Task
+  in a Pipeline as the param of another Task in the pipeline which has a string param
+  * If there is no value for the specified key, the Pipeline execution should fail
+    (or we may be able to use [TEP-0048](https://github.com/tektoncd/community/pull/240)
+    to specify a default value to use in that case instead)
+
+## Proposal
+
+* We would add support for object types for results and params, in addition to the existing string and array support
+  * Initially we would only support string values, eventually we can expand this to all values (string, array, object)
+    (note that this is the case for arrays as well, we don't yet support arrays of arrays)
+  * Only string key types would be supported (which is
+    [the only key type supported by json](https://datatracker.ietf.org/doc/html/rfc7159#section-4))
```

Comment:
How do you plan to handle addressing keys that currently exceed the supported character set of Tekton, but are perfectly legitimate JSON keys?

For example, this partial schema is used by logstream to parse Knative controller logs:
```go
	var line struct {
		...
		Controller string    `json:"knative.dev/controller"`
		...
		Key        string    `json:"knative.dev/key"`
		...
```

This could easily be:
```yaml
         name: line
         type: object
         properties:
           "knative.dev/controller": { type: string}
           "knative.dev/key": { type: string}
           ...
```

You'd see similar struct annotations to parse out something like this (were it JSON instead of toml): https://github.com/tektoncd/community/pull/504/files#diff-3a0e781137a6fde42cba2d980801d772725cf4a45490ce8929ad5a9d81953401R314-R321

Reply:
quick answer is i think we'd eventually expand the supported character set